warp10/src/main/java/io/warp10/script/formatted/FormattedWarpScriptFunction.java

```diff
@@ -0,0 +1,305 @@
+   *
+   * @param stack
+   * @return stack
+   * @throws WarpScriptException
+   */
+  @Override
+  final public Object apply(WarpScriptStack stack) throws WarpScriptException {
+
+    List<ArgumentSpecification> args = getArguments().args;
+    List<ArgumentSpecification> optArgs = getArguments().optArgs;
+
+    //
+    // Sanity checks
+    //
+
+    if (null == args) {
+      throw new WarpScriptException(getClass().getCanonicalName() + "'s method getArguments() returned null. If no " +
+        "argument is expected, it should return an empty array instead.");
+    }
+
+    if (null == optArgs) {
+      throw new WarpScriptException(getClass().getCanonicalName() + "'s method getOptionalArguments() returned null." +
+        " If no argument is expected, it should return an empty array instead.");
+    }
+
+    for (ArgumentSpecification arg: args) {
+      if (arg.isOptional()) {
+        throw new IllegalStateException("Output of " + getClass().getCanonicalName() + "'s method getArguments() must" +
+          " only contain arguments without a default value.");
+      }
+    }
+
+    for (ArgumentSpecification arg: optArgs) {
+      if (!arg.isOptional()) {
+        throw new IllegalStateException("Output of " + getClass().getCanonicalName() + "'s method getArguments() must" +
+          " only contain arguments with a default value.");
+      }
+    }
+
+    //
+    // If args and opt args are empty, should expect nothing on top
+    // If args is empty but opt args is not, should expect a map (possibly empty) on top
+    //
+
+    if (0 == args.size() && 0 == optArgs.size()) {
+      return apply(new HashMap<String, Object>(), stack);
+    }
+
+    if (0 == args.size()) {
+
+      if (0 == stack.depth() || !(stack.peek() instanceof Map)) {
+        throw new WarpScriptException(getClass().getCanonicalName() + " expects a MAP on top of the stack. To use default argument values, an empty MAP is expected.");
+      }
+    }
+
+    //
+    // If there are possible optional arguments, check that last mandatory argument is not a Map so that there is no
+    // confusion
+    //
+
+    if (args.size() > 1 && args.get(args.size() - 1) instanceof Map && optArgs.size() > 0) {
+      throw new IllegalStateException("In this case, the last non-optional argument can not be a Map so as to avoid any confusion when using optional arguments.");
+    }
+
+    //
+    // Extract arguments off the top of the stack
+    //
+
+    Map<String, Object> formattedArgs;
+
+    if (stack.peek() instanceof Map) {
+
+      //
+      // Case 1: A map is on top of the stack (some optional arguments may be given)
+      //
+
+      Map<String, Object> map = (Map) stack.peek();
+
+      //
+      // Check that non-optional args are contained in the map and that they have the correct type
+      //
+
+      for (ArgumentSpecification arg: args) {
+        if (!map.containsKey(arg.getName())) {
```

Comment:
It is better to call map.get() and check that the returned value is not null, containsKey is more costly, and in case it is not null, you are using the returned value afterwards, you could save one call.